// @smartthings/core-sdk/dist/endpoint/channels.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelsEndpoint = void 0;
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = require("../endpoint-client");
class ChannelsEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.EndpointClient('distchannels', config));
    }
    async create(data) {
        return this.client.post('', data);
    }
    async delete(id) {
        await this.client.delete(id);
    }
    async update(id, data) {
        return this.client.put(id, data);
    }
    async get(id) {
        return this.client.get(id);
    }
    async getDriverChannelMetaInfo(channelId, driverId) {
        return this.client.get(`${channelId}/drivers/${driverId}/meta`);
    }
    async list(options = {}) {
        const params = {};
        if (options.subscriberType) {
            params.type = options.subscriberType;
        }
        if (options.subscriberId) {
            if (!options.subscriberType) {
                throw Error('specifying a subscriberId requires also specifying a subscriberType');
            }
            params.subscriberId = options.subscriberId;
        }
        if (typeof (options.includeReadOnly) === 'boolean') {
            params.includeReadOnly = options.includeReadOnly.toString();
        }
        return this.client.getPagedItems('', params);
    }
    async listAssignedDrivers(channelId) {
        return this.client.getPagedItems(`${channelId}/drivers`);
    }
    /**
     * Assign or publish a driver to a channel.
     *
     * NOTE: This method also works to update the driver version assigned to a channel.
     */
    async assignDriver(channelId, driverId, version) {
        return this.client.post(`${channelId}/drivers`, { driverId, version });
    }
    async unassignDriver(channelId, driverId) {
        await this.client.delete(`${channelId}/drivers/${driverId}`);
    }
    async enrollHub(channelId, hubId) {
        await this.client.post(`${channelId}/hubs/${hubId}`);
    }
    async unenrollHub(channelId, hubId) {
        await this.client.delete(`${channelId}/hubs/${hubId}`);
    }
}
exports.ChannelsEndpoint = ChannelsEndpoint;
