// @smartthings/core-sdk/dist/endpoint/hubdevices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HubdevicesEndpoint = void 0;
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = require("../endpoint-client");
class HubdevicesEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.EndpointClient('hubdevices', config));
    }
    /**
     * Get a hub record
     * @param hubId
     */
    async get(hubId) {
        return this.client.get(`${hubId}`);
    }
    /**
     * Get the characteristics of a hub
     * @param hubId
     */
    async getCharacteristics(hubId) {
        return this.client.get(`${hubId}/characteristics`);
    }
    /**
     * Install driver on a hub. The primary use case of this functionality is to install a
     * self-published driver to be included in generic discovery (e.g. scanning).
     */
    async installDriver(driverId, hubId, channelId) {
        return this.client.put(`${hubId}/drivers/${driverId}`, { channelId });
    }
    /**
     * Change the driver for a device to the one specified by driverId.
     */
    async switchDriver(driverId, hubId, deviceId, forceUpdate = false) {
        return this.client.patch(`${hubId}/childdevice/${deviceId}`, { driverId }, forceUpdate ? { forceUpdate: 'true' } : undefined);
    }
    /**
     * Removes a driver from being installed on a hub. This will allow the hub to clean up the
     * deleted driver. However, all dependent devices need to be removed for cleanup to fully occur.
     */
    async uninstallDriver(driverId, hubId) {
        return this.client.delete(`${hubId}/drivers/${driverId}`);
    }
    /**
     * List drivers installed on the hub.
     *
     * @param deviceId When included, limit the drivers to those marked as matching the specified device.
     */
    async listInstalled(hubId, deviceId) {
        const params = deviceId ? { deviceId } : undefined;
        return this.client.get(`${hubId}/drivers`, params);
    }
    async getInstalled(hubId, driverId) {
        return this.client.get(`${hubId}/drivers/${driverId}`);
    }
    /**
     * Returns the list of driver channels the hub is currently subscribed to.
     * Currently only returns the driver channel type.
     */
    async enrolledChannels(hubId) {
        return this.client.get(`${hubId}/channels`, { channelType: 'DRIVERS' });
    }
}
exports.HubdevicesEndpoint = HubdevicesEndpoint;
