// @smartthings/core-sdk/dist/endpoint/notifications.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationsEndpoint = exports.DeepLinkType = exports.NotificationRequestType = void 0;
const endpoint_1 = require("../endpoint");
const endpoint_client_1 = require("../endpoint-client");
var NotificationRequestType;
(function (NotificationRequestType) {
    NotificationRequestType["ALERT"] = "ALERT";
    NotificationRequestType["SUGGESTED_ACTION"] = "SUGGESTED_ACTION";
    NotificationRequestType["EVENT_LOGGING"] = "EVENT_LOGGING";
    NotificationRequestType["AUTOMATION_INFO"] = "AUTOMATION_INFO";
})(NotificationRequestType || (exports.NotificationRequestType = NotificationRequestType = {}));
var DeepLinkType;
(function (DeepLinkType) {
    DeepLinkType["device"] = "device";
    DeepLinkType["installedApp"] = "installedApp";
    DeepLinkType["location"] = "location";
})(DeepLinkType || (exports.DeepLinkType = DeepLinkType = {}));
class NotificationsEndpoint extends endpoint_1.Endpoint {
    constructor(config) {
        super(new endpoint_client_1.EndpointClient('notification', config));
    }
    /**
     * Sends a push notification to mobile apps belonging to the location
     * @param data the notification request. If the client has been configured with a location ID it can be omitted
     * from this request.
     */
    async create(data) {
        data.locationId = this.locationId(data.locationId);
        return this.client.post(undefined, data);
    }
}
exports.NotificationsEndpoint = NotificationsEndpoint;
